project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/contact/OrcidRepository.java

```diff
@@ -0,0 +1,230 @@
+  }
+
+  private static HttpClient createHttpClient() {
+    return HttpClient.newBuilder().version(HttpClient.Version.HTTP_2)
+        .followRedirects(HttpClient.Redirect.NORMAL).connectTimeout(
+            Duration.ofSeconds(10)).build();
+  }
+
+  private static AuthResponse authenticate(HttpClient client, String clientID, String clientSecret,
+      String tokenEndpoint) {
+    var params = Map.of("client_id", clientID, "client_secret", clientSecret, "scope",
+        "/read-public", "grant_type", "client_credentials");
+
+    String form = params.entrySet().stream()
+        .map(entry -> URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8) + "=" +
+            URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8))
+        .collect(Collectors.joining("&"));
+
+    // Build the HTTP request
+    HttpRequest request = HttpRequest.newBuilder()
+        .uri(URI.create(tokenEndpoint))
+        .header("Content-Type", "application/x-www-form-urlencoded")
+        .POST(HttpRequest.BodyPublishers.ofString(form))
+        .build();
+
+    try {
+      var response = client.send(request, HttpResponse.BodyHandlers.ofString());
+      ObjectMapper mapper = new ObjectMapper();
+      try (var parser = mapper.createParser(response.body())) {
+        return parser.readValueAs(AuthResponse.class);
+      }
+    } catch (IOException | InterruptedException e) {
+      if (e instanceof InterruptedException) {
+        Thread.currentThread().interrupt();
+      }
+      log.error("Error sending orcid request", e);
+      throw new OrcidRepository.QueryException("Authentication failed", e);
+    }
+
+  }
+
+  private static OrcidEntry convert(OrcidRecord orcidRecord) {
+    var emailList = Arrays.stream(orcidRecord.email()).toList();
+    //We want to show the full name meaning first and last name in a human readable format, ensuring that no trailing whitespaces are added
+    var fullName = orcidRecord.givenName() + " " + orcidRecord.familyName().trim();
+    var orcid = orcidRecord.orcidID();
+    //If an orcid record does not contain a name or email it is considered invalid and will not be considered further
+    if (orcid.isBlank()) {
+      return null;
+    }
+    if (fullName.isBlank()) {
+      return null;
+    }
+    if (emailList.isEmpty()) {
+      return null;
+    }
+    var email = emailList.stream().findFirst().orElse("");
+    return new OrcidEntry(fullName, email, orcid, OIDC_ISSUER);
+  }
+
+  @Override
+  public List<OrcidEntry> findAll(String query, int limit, int offset) {
+    String url = paginatedQuery;
+    url += "+AND+email:*"; //require an email to be present
+    url += "+AND+given-names:*"; //require a first name to be present
+    url += "+AND+family-name:*"; //require a family name to be present
+
+    //We want to enable the user to enter a query with spaces (e.g. searching for John Doe), so we need to sanitize the query string
+    String sanitizedQuery = query.trim().replaceAll("\\s+", "+AND+");
```

Comment:
I am not sure that this works reliably. Why not use the URLEncoder designed for this purpose?

Reply:
That's a valid point i adapted it 👍